File: HADES/FEED SIZING/SonicNozzleSizing.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%   Purpose: Sonic Nozzle Sizing from Downstream
%            Pressure and Mass Flow Requirements
%
%   Programmers: Deepesh Balwani, Noah Ha
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

clc;
clear;

%% CONVERSIONS

in_to_m   = 0.0254;
lb_to_kg  = 0.453592;
psi_to_Pa = 6894.76;

%% DESIGN REQUIREMENTS

% Mass flow requirement
m_dot_req   = 2 * lb_to_kg;      % Required mass flow [kg/s]
flow_margin = 1.15;              % 15% design margin
m_dot       = m_dot_req * flow_margin;

% Downstream pressure requirement
P_down = 20e5;                   % Downstream/back pressure [Pa] = 20 bar

% Air properties
R       = 287;                   % Specific gas constant [J/(kg*K)]
gamma   = 1.4;                   % Specific heat ratio [-]
T_tu    = 283;                   % Upstream total temperature [K]

%% CHOKING CONDITION

% Critical pressure ratio for choking:
%
% P_t / P_tu = (2/(gamma+1))^(gamma/(gamma-1))
%
% For air, this is approximately 0.5283.

crit_ratio = (2/(gamma + 1))^(gamma/(gamma - 1));

% Minimum upstream total pressure required for choking.
%
% At the limiting choking condition:
%
% P_down = P_t = crit_ratio * P_tu
%
% Therefore:
%
% P_tu_min = P_down / crit_ratio

P_tu_min = P_down / crit_ratio;

%% CHOKED FLOW FUNCTION

% Choked mass-flow parameter:
%
% mdot = (P_tu * A_t / sqrt(T_tu)) *
%        sqrt(gamma/R) *
%        (2/(gamma+1))^((gamma+1)/(2*(gamma-1)))

choked_param = sqrt(gamma / R) * ...
               (2 / (gamma + 1))^((gamma + 1) / ...
               (2 * (gamma - 1)));

%% THROAT AREA

% Rearranging the choked-flow equation:
%
% A_t = mdot * sqrt(T_tu) / (P_tu * choked_param)

A_t = (m_dot * sqrt(T_tu)) / ...
      (P_tu_min * choked_param);

%% THROAT DIAMETER

D_t_m = 2 * sqrt(A_t / pi);

%% THROAT CONDITIONS

% At M = 1:
%
% T_t / T_tu = 2/(gamma+1)
% P_t / P_tu = crit_ratio
%
% Calculate throat static conditions.

T_throat = T_tu * (2 / (gamma + 1));
P_throat = P_tu_min * crit_ratio;

%% FEED LINE GEOMETRY

wall_thickness = 0.049 * in_to_m;

feed_OD = (16 / 16) * in_to_m;       % 1.000 in OD
feed_ID = feed_OD - 2 * wall_thickness;

%% OUTPUTS

fprintf('\n');
fprintf('=============================================\n');
fprintf('       SONIC NOZZLE SIZING RESULTS\n');
fprintf('=============================================\n');

fprintf('\n--- DESIGN REQUIREMENTS ---\n');

fprintf('Required Mass Flow          : %.3f lb/s\n', ...
        m_dot_req / lb_to_kg);

fprintf('Design Mass Flow            : %.3f lb/s\n', ...
        m_dot / lb_to_kg);

fprintf('Design Mass Flow            : %.3f kg/s\n', ...
        m_dot);

fprintf('Downstream Pressure         : %.2f psi\n', ...
        P_down / psi_to_Pa);

fprintf('Downstream Pressure         : %.2f bar\n', ...
        P_down / 1e5);

fprintf('Upstream Total Temperature  : %.2f K\n', ...
        T_tu);

fprintf('\n--- CHOKING CONDITION ---\n');

fprintf('Critical Pressure Ratio     : %.4f\n', ...
        crit_ratio);

fprintf('Minimum Upstream Total P    : %.2f psi\n', ...
        P_tu_min / psi_to_Pa);

fprintf('Minimum Upstream Total P    : %.2f bar\n', ...
        P_tu_min / 1e5);

fprintf('\n--- THROAT ---\n');

fprintf('Throat Area                 : %.6f m^2\n', ...
        A_t);

fprintf('Throat Area                 : %.4f in^2\n', ...
        A_t / in_to_m^2);

fprintf('Throat Diameter             : %.2f thou\n', ...
        (D_t_m / in_to_m) * 1000);

fprintf('Throat Diameter             : %.4f in\n', ...
        D_t_m / in_to_m);

fprintf('Throat Diameter             : %.3f mm\n', ...
        D_t_m * 1000);

fprintf('\n--- THROAT STATIC CONDITIONS ---\n');

fprintf('Throat Pressure (M=1)       : %.2f psi\n', ...
        P_throat / psi_to_Pa);

fprintf('Throat Pressure (M=1)       : %.2f bar\n', ...
        P_throat / 1e5);

fprintf('Throat Temperature (M=1)    : %.2f K\n', ...
        T_throat);

fprintf('\n--- FEED LINE ---\n');

fprintf('Feed OD                     : %.3f in\n', ...
        feed_OD / in_to_m);

fprintf('Feed ID                     : %.3f in\n', ...
        feed_ID / in_to_m);

fprintf('=============================================\n');

File: HADES/FEED SIZING/SonicNozzleSizing.py
import math

# Conversions
in_to_m = 0.0254
lb_to_kg = 0.453592

# Feed requirements
m_dot = 3 * lb_to_kg # kg/s
P_m = 2 * 10**6 # Pa
R = 287 # J/kg-K
gamma = 1.4
T = 283 # K
mass_flow_margin = 0.85
crit_ratio = (2 / (gamma+1))**(gamma/(gamma-1))
P_t_u_req = P_m / crit_ratio

# Solve for the feed area
wall_thickness = 0.095
feed_diameter = (16/16 - wall_thickness*2) * in_to_m
A_2 = math.pi * feed_diameter**2 / 4

# Solve for the downstream Mach number required
M_2 = (m_dot * mass_flow_margin) / (P_m * A_2) * math.sqrt(R * T / gamma)

# Solve for throat area
A_factor = 2/(gamma+1) * (1 + (gamma-1)/2 * M_2**2) ** (0.5 * (gamma+1)/(gamma-1))
A_t = A_2 / A_factor * M_2
D_t = 2 * math.sqrt(A_t/math.pi)

# Solve for total upstream pressure
T_t = T * (1 + (gamma-1)/2 * M_2**2)
A_1 = A_2
P_t_u = m_dot/A_t * math.sqrt(T_t * R / gamma) * ((gamma+1)/2)**((gamma+1)/(2*(gamma-1)))

# Output statements
print()
print("M_2 =", str(M_2))
print("A_2 =", str(A_2), "m^2")
print("A_t =", str(A_t), "m^2")
print("Throat Diameter:", str(D_t * 1000), "mm")
print("Upstream Pressure:", str(P_t_u * 10**(-5)), "bar")
print("Required Upstream Pressure:", str(P_t_u_req * 10**(-5)), "bar")

# Print if it chokes
if P_t_u > P_t_u_req:
    print("Flow is choked!")
else:
    print("Flow does not choke :(")

print()